helix-lsp/src/lib.rs

```diff
@@ -343,52 +343,81 @@ impl Registry {
             None => return Err(Error::LspNotDefined),
         };
 
-        match self.inner.entry(language_config.scope.clone()) {
-            Entry::Occupied(entry) => Ok(entry.get().1.clone()),
-            Entry::Vacant(entry) => {
-                // initialize a new client
-                let id = self.counter.fetch_add(1, Ordering::Relaxed);
-                let (client, incoming, initialize_notify) = Client::start(
-                    &config.command,
-                    &config.args,
-                    language_config.config.clone(),
-                    &language_config.roots,
-                    id,
-                )?;
-                self.incoming.push(UnboundedReceiverStream::new(incoming));
-                let client = Arc::new(client);
-
-                // Initialize the client asynchronously
-                let _client = client.clone();
-                tokio::spawn(async move {
-                    use futures_util::TryFutureExt;
-                    let value = _client
-                        .capabilities
-                        .get_or_try_init(|| {
-                            _client
-                                .initialize()
-                                .map_ok(|response| response.capabilities)
-                        })
-                        .await;
-
-                    if let Err(e) = value {
-                        log::error!("failed to initialize language server: {}", e);
-                        return;
-                    }
-
-                    // next up, notify<initialized>
-                    _client
-                        .notify::<lsp::notification::Initialized>(lsp::InitializedParams {})
-                        .await
-                        .unwrap();
+        if let Some((_, client)) = self.inner.get(&language_config.scope) {
+            Ok(client.clone())
+        } else {
+            let id = self.counter.fetch_add(1, Ordering::Relaxed);
+            let client = self.initialize_client(language_config, config, id)?; // initialize a new client
+            self.inner
+                .insert(language_config.scope.clone(), (id, client.clone()));
+            Ok(client)
+        }
+    }
```

Comment:
Why the change here? This was deliberate since we don't want to try reinitializing the client if we already tried starting it and failed.

Reply:
the functionality of `get` should be unchanged; I pulled out `initialize_client` into its own function so that I could also use it in `restart`. I switched from using `self.inner.entry` to `self.inner.get` to avoid multiple mutable borrows.